pkg/sync/syncWorker.go

```diff
@@ -318,55 +336,62 @@ func (s *syncWorker) listenToStream(
 		}
 		s.log.Debug("Received envelopes", zap.Any("numEnvelopes", len(envs.Envelopes)))
 		for _, env := range envs.Envelopes {
-			s.insertEnvelope(env)
+			s.validateAndInsertEnvelope(stream, env)
 		}
 	}
-
 }
 
-func (s *syncWorker) insertEnvelope(env *envelopes.OriginatorEnvelope) {
-	s.log.Debug("Replication server received envelope", zap.Any("envelope", env))
-	// TODO(nm) Validation logic - share code with API service and publish worker
-	originatorBytes, err := proto.Marshal(env)
+func (s *syncWorker) validateAndInsertEnvelope(
+	stream *originatorStream,
+	envProto *envelopes.OriginatorEnvelope,
+) {
+	env, err := envUtils.NewOriginatorEnvelope(envProto)
 	if err != nil {
-		s.log.Error("Failed to marshal originator envelope", zap.Error(err))
+		s.log.Error("Failed to unmarshal originator envelope", zap.Error(err))
 		return
 	}
 
-	unsignedEnvelope := &envelopes.UnsignedOriginatorEnvelope{}
-	err = proto.Unmarshal(env.GetUnsignedOriginatorEnvelope(), unsignedEnvelope)
-	if err != nil {
-		s.log.Error(
-			"Failed to unmarshal unsigned originator envelope",
-			zap.Error(err),
-		)
+	if env.OriginatorNodeID() != stream.nodeID {
+		s.log.Error("Received envelope from wrong node", zap.Any("nodeID", env.OriginatorNodeID()))
 		return
 	}
 
-	clientEnvelope := &envelopes.ClientEnvelope{}
-	err = proto.Unmarshal(
-		unsignedEnvelope.GetPayerEnvelope().GetUnsignedClientEnvelope(),
-		clientEnvelope,
-	)
+	var lastSequenceID uint64 = 0
+	var lastNs int64 = 0
+	if stream.lastEnvelope != nil {
+		lastSequenceID = stream.lastEnvelope.OriginatorSequenceID()
+		lastNs = stream.lastEnvelope.OriginatorNs()
+	}
+	if env.OriginatorSequenceID() != lastSequenceID+1 || env.OriginatorNs() < lastNs {
+		// TODO(rich) Submit misbehavior report and continue
+		s.log.Error("Received out of order envelope")
+	}
+
+	if env.OriginatorSequenceID() > lastSequenceID {
+		stream.lastEnvelope = env
+	}
```

Comment:
do we check this anywhere at stream startup? What if we rebuild the stream after network failure?

Reply:
If you look at where the `originatorStream` is constructed, we are fetching the last payload out of the database before stream construction